File: Arrays/347. Top K Frequent Elements/347. Top K Frequent Elements.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var topKFrequent = function (nums, k) {
  let counter = {};
  for (const n of nums) {
    counter[n] = (counter[n] || 0) + 1;
  }

  let arr = new Array(nums.length + 1).fill(0);
  for (const [num, freq] of Object.entries(counter)) {
    if (arr[freq] === 0) arr[freq] = [Number(num)];
    else arr[freq].push(Number(num));
  }

  let ans = [];
  for (let i = nums.length; i > 0; i--) {
    if (arr[i] !== 0) ans.push(...arr[i]);
    if (ans.length >= k) break;
  }

  return ans;
};

// Time: O(n)
// Space: O(n)

File: Arrays/347. Top K Frequent Elements/347. Top K Frequent Elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        counter = Counter(nums)
        buckets = [0] * (len(nums) + 1)

        for n, count in counter.items():
            if buckets[count] == 0:
                buckets[count] = [n]
            else:
                buckets[count].append(n)

        ans = []
        for i in range(len(nums), -1, -1):
            if buckets[i] != 0:
                ans += buckets[i]
            if len(ans) == k:
                return ans

# Time: O(n)
# Space: O(n)
